clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java

```diff
@@ -652,7 +655,7 @@ public void subscribe(List<String> topics, ConsumerRebalanceListener listener) {
         try {
             log.debug("Subscribed to topic(s): {}", Utils.join(topics, ", "));
             this.subscriptions.subscribe(topics, listener);
-            metadata.setTopics(topics);
+            metadata.setTopics(subscriptions.groupSubscription());
```

Comment:
The leader is the only member who sees the group subscription; for everyone else, groupSubscription() will just return that member's subscription. We depend on the leader's metadata in order to set assignments, so it must have metadata available for the full group's subscription. We could just issue one metadata request when the leader is preparing to do the assignment, but that opens the door to the leader missing metadata changes affecting topics it itself is not subscribed to. More concretely, suppose the leader is subscribed to [a], but one follower is subscribed to [a,b]. If the follower notices an increase in b's partitions, it can trigger a rebalance, but there's no guarantee that the partition change will be visible by the leader when it fetches topic metadata. Therefore we register the leader to watch for changes to b as well. Eventually, when the leader sees the change, it can rebalance. Does that make sense?